src/iterable-request.ts

```diff
@@ -0,0 +1,254 @@
+    if (this.terminating) {
+      if (this.terminatorResolve) {
+        this.terminatorResolve();
+      }
+      return;
+    }
+    if (error && !this.error) {
+      this.error = error;
+    }
+    this.servePromise();
+  }
+
+  private columnMetadataEventHandler = (columnMetadata: ColumnMetadata[] | Record<string, ColumnMetadata>) => {
+    this.resultSetNo++;
+    this.columnMetadata = columnMetadata;
+  };
+
+  private rowEventHandler = (row: RowData) => {
+    if (this.requestCompleted || this.error || this.terminating) {
+      return;
+    }
+    if (this.resultSetNo === 0 || !this.columnMetadata) {
+      this.error = new Error('No columnMetadata event received before row event.');
+      this.servePromise();
+      return;
+    }
+    const item: IterableRequestItem = { row, resultSetNo: this.resultSetNo, columnMetadata: this.columnMetadata };
+    this.fifo.push(item);
+    if (this.fifo.length >= this.fifoPauseLevel && !this.requestPaused) {
+      this.request.pause();
+      this.requestPaused = true;
+    }
+    this.servePromise();
+  };
+
+}
+
+// Internal class for the iterator object which is passed to the API client.
+class IterableRequestIterator implements AsyncIterator<IterableRequestItem> {
+
+  private controller: IterableRequestController;
+
+  constructor(controller: IterableRequestController) {
+    this.controller = controller;
+  }
+
+  public next(): Promise<IteratorResult<IterableRequestItem>> {
+    return new Promise<IteratorResult<IterableRequestItem>>(this.controller.promiseExecutor);
+  }
+
+  public async return(value?: any): Promise<any> {
+    await this.controller.terminate();
+    return Promise.resolve({ value, done: true });         // eslint-disable-line @typescript-eslint/return-await
+  }
+
+  public async throw(_exception?: any): Promise<any> {
+    await this.controller.terminate();
+    return Promise.resolve({ done: true });                // eslint-disable-line @typescript-eslint/return-await
+  }
+
+}
+
+/**
+* An iterable `Request` class.
+*
+* This iterable version is a super class of the normal `Request` class.
+*
+* Usage:
+* ```js
+* const request = new IterableRequest("select 42, 'hello world'");
+* connection.execSql(request);
+* for await (const item of request) {
+*   console.log(item.row);
+* }
+* ```
+*/
+class IterableRequest extends Request implements AsyncIterable<IterableRequestItem> {
+
+  private iterator: IterableRequestIterator;
+
+  constructor(sqlTextOrProcedure: string | undefined, options?: IterableRequestOptions) {
+    super(sqlTextOrProcedure, completionCallback, options);
+    const controller = new IterableRequestController(this, options);
+    this.iterator = new IterableRequestIterator(controller);
+
+    function completionCallback(error: Error | null | undefined) {
+      if (controller) {
+        controller.completionCallback(error);
+      }
+    }
+  }
+
+  [Symbol.asyncIterator](): AsyncIterator<IterableRequestItem> {
+    return this.iterator;
+  }
```

Comment:
Have you thought about implementing this through using `events.on`? I haven't tested this, but it would allow you to get rid of the custom iterator implementation.

```
  constructor(sqlTextOrProcedure: string | undefined, options?: IterableRequestOptions) {
    super(sqlTextOrProcedure, completionCallback, options);
    const controller = new AbortController();
    this.doneSignal = controller.signal;

    function completionCallback(error: Error | null | undefined) {
       controller.abort(error);
    }
  }

  [Symbol.asyncIterator](): AsyncIterator<IterableRequestItem> {
    return events.on(this, 'row', { signal: this.doneSignal });
  }
```

Reply:
Thanks. I didn't know about `events.on()`. It does more or less the same as my controller and iterator classes.
I will refactor my code into using `events.on()`.

Reply:
Since Tedious still has to support Node 18, I had a look at that [old implementation of `event.on()`](https://github.com/nodejs/node/blob/e578c0b1e8d3dd817e692a0c5df1b97580bc7c7f/lib/events.js#L1046).
The `close` event was not yet implemented in that version. I can't find a proper way to end the iteration so that the remaining items in the FIFO are still processed. `abortListener()` calls `errorHandler()` which calls `toError.reject(err)` which rejects the waiting `Promise`, 
Calling `iterator.return()` resolves the waiting `Promise` with `done = true`.

Should I use my implementation for old Node versions and `event.on()` for new Node versions?